File: combinationSum.py
def is_valid_state(state, target):
    if( sum( state ) == target):
        return True
    return False

def get_candidates(state):
    return []

def search(state, solutions):
    if is_valid_state(state):
        solutions.append(state.copy())
        # return

    for candidate in get_candidates(state):
        state.add(candidate)
        search(state, solutions)
        state.remove(candidate)

def solve():
    solutions = []
    state = set()
    search(state, solutions)
    return solutions

File: combinationSum.js


function combinationSum( candidates, target, sum=0, index=0, state=[], result=[] ){
    if( sum === target ){
        return result.push( state.slice() );
    }else if( sum > target ){
        return;
    }else{
        for( let i = index; i < candidates.length; i++ ){
            state.push( candidates[ i ] );
            combinationSum( candidates, target, sum + candidates[ i ], i, state, result );
            state.pop();
        }
    }
    return result;
}

console.log( combinationSum( [1,2,3,4], 5 ))